docs/specs/fee-calculator.md

```diff
@@ -0,0 +1,405 @@
+
+### Edge Case Handling
+
+1. **First Day**: No previous balance, so balance change = 0
+2. **Missing Events**: If no events for a date, distributions = 0
+3. **Data Gaps**: Report gaps but continue processing available data
+4. **Zero Balances**: Valid scenario, process normally
+
+## 7. Data Structures
+
+### Input Types
+
+```typescript
+interface DistributorsData {
+  distributors: Array<{
+    address: string;
+    metadata: {
+      name: string;
+      created_at: string;
+    };
+  }>;
+}
+
+interface BalanceData {
+  metadata: {
+    chain_id: number;
+    reward_distributor: string;
+  };
+  balances: {
+    [date: string]: {
+      block_number: number;
+      balance_wei: string;
+    };
+  };
+}
+
+interface RecipientRecievedEventData {
+  metadata: {
+    chain_id: number;
+    distributor_address: string;
+  };
+  events: Array<{
+    block_number: number;
+    transaction_hash: string;
+    log_index: number;
+    timestamp: string;
+    date: string;
+    recipient: string;
+    amount_wei: string;
+  }>;
+}
+```
+
+### Output Types
+
+```typescript
+interface FeeReport {
+  metadata: {
+    chain_id: number;
+    generated_at: string;
+    report_type: "daily_fees";
+  };
+  distributors: {
+    [distributorAddress: string]: {
+      name: string;
+      daily_fees: Array<{
+        date: string;
+        start_balance_wei: string;
+        end_balance_wei: string;
+        balance_change_wei: string;
+        distributions_wei: string;
+        distributions_count: number;
+        fee_wei: string;
+      }>;
+    };
+  };
+}
+```
+
+### Internal Calculation Types
+
+```typescript
+interface DailyFeeCalculation {
+  date: string;
+  startBalance: bigint;
+  endBalance: bigint;
+  balanceChange: bigint;
+  distributions: bigint;
+  distributionCount: number;
+  totalFee: bigint;
+}
+```
+
+## 8. Implementation Requirements
+
+### Validation Rules
+
+1. **Address Validation**:
+
+   - All addresses must be valid Ethereum addresses
+   - Addresses normalized to checksummed format
+
+2. **Date Validation**:
+
+   - All dates in ISO format (YYYY-MM-DD)
+   - Dates must be chronologically ordered
+   - Cannot calculate fees for future dates
+
+3. **Amount Validation**:
+
+   - All amounts must be valid numeric strings
+   - Handle large numbers using bigint internally
+   - Convert back to strings for output
+
+4. **Data Completeness**:
+   - Log warnings for missing data but continue processing
+   - Clearly indicate data gaps in output
+
+### Security Considerations
+
+1. **Data Integrity**:
+
+   - Validate all numeric values before processing
+   - Use bigint to prevent overflow with large wei values
+   - FileManager ensures atomic file operations
+
+2. **Error Handling**:
+   - Never expose system paths in error messages
+   - Fail gracefully with descriptive error messages
+
+## 9. Test Data Requirements
+
+### Minimum Test Data
+
+Integration tests require real blockchain data from Arbitrum Nova:
```

Comment:
Take a look in the test data folder in the tests directory. Have a look through the test data that's there. Right below where you think that test data is available or not. If you don't think it's available, write what needs to be fetched or generated or populated.